app/routes.py

```diff
@@ -0,0 +1,49 @@
+from io import BytesIO
+import os.path
+from flask import Response, make_response, jsonify, abort
+from PIL import Image
+from app import app
+from app.helpers.check_functions import check_color_channels
+from app.helpers.route import prefix_route
+from app.helpers import make_error_msg
+
+# add route prefix
+app.route = prefix_route(app.route, '/v4/color/')
```

Comment:
Should the versioning not handle by `traefik`?

Reply:
> Looks very appart the following three points:
> 
> * Origin check in before_request (see related comment)
> * Change the default port from 8080 to 5000 in Makefile and in wsgy.py (Docker should be left to 8080)
> * Logging

first two bullets are implemented
Logging will be done in the next iteration, together with the testing (working on it already).

Reply:
I'll leave that for discussion. Would be also of relevance for service-qrcode et al.

Reply:
Good point, In case where we really need to operate several versions in parallel it'll either be handled by traefik or directly cloudfront. If it's the same service handling several versions we can think about a pattern like this (dunno if it's possible with prefix_route but you'll get the idea):
```
@app.route('/<string:version>/color/...`)
def foo(version,....):
    if version == 'v1':
        # handle according to v1 
    elif version == 'v2':
        # handle accroding to v2
    else:
        raise UnsupportedVersionError
```
What I would avoid is doing path rewrite in CF and traefik

Reply:
Simple version handling is now implemented in `app.route = prefix_route(app.route, '/v<int:ver>/color')`. Versions are supposed to be integers, is that OK? And as I do not know about versions 1 to 3, simply all version <= 4 are accepted, and for all versions >4 an "unsupported version" error is raised.

I guess there are more elegant solutions for this, but if it will be handled by traefik or cloudfront, I guess this simple approach might do for now.

Reply:
The way to handle prefixes in flask is with environ['SCRIPT_NAME'] and ['PATH_INFO'] which are actually used within the middleware "reverseproxy". If we want to add a prefix within the code, it should be done here.

If we want to support multiple versions of the same service for some time (I guess this would start with v5), I would not bloat the code with multiple versions and, instead, would deploy two instances (one for each version) and let `traefik` take care of the correspondence between routes and correct service.

That way, we can announce to our users that v4 is deprecated, publish v5, and give them some time (three months seems good to basically change a version number in a route) before decommissioning the v4. This would also avoid maintaining the v4 when the v5/6/7 is out, or mixing versions together within the code.

Reply:
I like the idea of having multiple service version in separate instance and letting traefik handling the routing of the version

Reply:
OK, so if I understand you (and `middleware.py`) correctly, then this could be the place to change the route and add version info: `app.wsgi_app = ReverseProxy(app.wsgi_app, script_name='/')` inside the apps `__init__.py` by adapting the script_name, right?
But then you and also Marc suggest to to the version handling outside the code handled by `traefik`. Guess we need to make up our "common mind" about this, so we can proceed here and also with the other microservices currently in progress. Maybe we can have a short discussion today during or after the daily?

Reply:
The first part of my message was mostly in "where prefixing should happen when it's within the code", but in our current case, `traefik` can do all the work for us and we won't even need to prefix the routes.